## How to Write Tests

The bodies of test functions typically perform these three actions:
1. Set up any needed data or state.
2. Run the code you want to test.
3. Assert the results are what you expect.

### The Anatomy of a Test Function
you run your tests with the `cargo test` command, and Rust then builds a test runner binary that runs the annotated functions and reports on whether each test function passes or fails.

Whenever we make a new library project with Cargo, a test module with a test function in it is automatically generated for us. This module gives you a template for writing your tests

```rust
#[cfg(test)]
mod tests {
    #[test]
    fn exploration() {
        let result = 2 + 2;
        assert_eq!(result, 4);
    }
}
```
- `#[test]` indicates this is a test function, so the test runner knows to treat this function as a test. But we can also have non test functions in the `tests` module
- `assert_eq!` macro compares the left and right arguments to check whether they are equal or not

To make a test fail, call the `panic!` macro:
```rust
#[cfg(test)]
mod tests {
    #[test]
    fn exploration() {
        assert_eq!(2 + 2, 4);
    }

    #[test]
    fn another() {
        panic!("Make this test fail");   // the "another()" test will fail b/c of panic!
    }
}
```

### Checking Results with the `assert!` Macro
The `assert!` macro is useful when you want to ensure that some condition in a test evaluates to `true`
- `assert!` takes an argument that evaluates to a bool

### Testing Equality with the `assert_eq!` and `assert_ne!` Macros
`asser_eq!` returns true if it's two arguments **ARE** equal, whereas `assert_ne!` returns true if it's two arguments **AREN'T** equal.
- Under the surface, the `assert_eq!` and `assert_ne!` macros use the operators `==` and `!=`, respectively

### Adding Custom Failure Messages
We can add a custom message to be printed with the failure message as an optional argument to the `assert!`, `assert_eq!`, and `assert_ne!` macros
- Any arguments specified after the required arguments are passed along to the `format!` macro, so you can pass a format string that contains `{}` placeholders and values to go in those placeholders

```rust
#[cfg(test)]
mod tests {
    #[test]
    fn greeting_contains_name() {
        let result = greeting("Carol");
        assert!( 
            result.contains("Carol"),
            "Greeting did not contain name, value was `{}`",    <==== custom failure message
            result        <==== parameter to pass into formatted string
        );
    }
}
```

### Checking for Panics with `should_panic`
`should_panic` is an attribute that can be added to a test function (like `#[test]`). The test function implementing `should_panic` passes if the code inside the function panics, and it fails if the code inside the function doesn’t panic:
```rust
pub struct Guess {
    value: i32,
}

impl Guess {
    pub fn new(value: i32) -> Guess {
        if value < 1 || value > 100 {
            panic!("Guess value must be between 1 and 100, got {}.", value);
        }

        Guess { value }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    #[should_panic]       <==== test to passes here because greater_than_100() does panic
    fn greater_than_100() {
        Guess::new(200);  <==== if this changed to a number in range (0, 100), then the test would fail
    }
}
```
- To make `should_panic` tests more precise, we can add an optional expected parameter to the` should_panic` attribute. The test harness will make sure that the failure message contains the provided text. For example:
    ```rust
    #[cfg(test)]
    mod tests {
        use super::*;

        #[test]
        #[should_panic(expected = "less than or equal to 100")]   
        fn greater_than_100() {
            Guess::new(200);
        }
    }
    ```

### Using `Result<T, E>` in Tests
Our tests so far all panic when they fail. We can also write tests that use `Result<T, E>`!:
```rust
#[cfg(test)]
mod tests {
    #[test]
    fn it_works() -> Result<(), String> {
        if 2 + 2 == 4 {
            Ok(())
        } else {
            Err(String::from("two plus two does not equal four"))
        }
    }
}
```
- rather than calling the `assert_eq!` macro, we return `Ok(())` when the test passes and an `Err` with a `String` inside when the test fails
- Writing tests so they return a `Result<T, E>` enables you to use the question mark operator (`?`) in the body of tests, which can be a convenient way to write tests that should fail if any operation within them returns an `Err` variant
- You can’t use the `#[should_panic]` annotation on tests that use `Result<T, E>`. To assert that an operation returns an `Err` variant, don’t use the question mark operator on the `Result<T, E>` value. Instead, use `assert!(value.is_err())`.

---

### Quiz
1. What is the annotation you add to a function to indicate that it's a unit test?
#### Answer:
`#[test]`

<br/>

2. Let's say you have a function with the type signature:
    ```rust
    fn f(x: usize) -> Result<usize, String>;
    ```
    And you want to test that `f(0)` should return `Err(_)`. Which of the following is NOT a valid way to test that?
#### Answer:
```rust
✅ CORRECT
#[test]
#[should_err]
fn test() -> Result<usize, String> {
    f(0)
}
```
`should_err` does not exist in Rust — any test that returns a `Result` must return `Ok` to pass.

---